odoo/api.py

```diff
@@ -559,7 +560,7 @@ def company(self):
                 if any(cid not in user_company_ids for cid in company_ids):
                     raise AccessError(_("Access to unauthorized or invalid companies."))
             return self['res.company'].browse(company_ids[0])
-        return self.user.company_id
+        return self.user.company_id.sudo(self.su)
```

Comment:
You can be more explicit about using the same env:
```suggestion
        return self.user.company_id.with_env(self)
```
Note that `sudo(self.su)` will instanciate an env, which searches for an env that matches the spec.

Reply:
Also document explicitly that `env.company` has `self` as `env`.